### PR TITLE
Nucleation data object to organize associated constants and Views

### PR DESCRIPTION
**Pull request by MattRolchigo**
Also rewrote unit tests to use the nucleation structure, otherwise this basically just moved existing functions into the new file CAnucleation.hpp

### CODE REVIEW

**Review by streeve — CHANGES_REQUESTED** [review]

File: src/runCA.cpp [hunk 1]
```diff
@@ -213,22 +214,18 @@ void RunProgram_Reduced(int id, int np, std::string InputFile) {
...
-
+    Nucleation<device_memory_space> nucleation(EstimatedNuclei_ThisRankThisLayer, NMax, deltax);
     // Fill in nucleation data structures, and assign nucleation undercooling values to potential nucleation events
     // Potential nucleation grains are only associated with liquid cells in layer 0 - they will be initialized for each
     // successive layer when layer 0 in complete
-    NucleiInit(0, RNGSeed, MyYSlices, MyYOffset, nx, ny, nzActive, ZBound_Low, id, NMax, dTN, dTsigma, deltax,
-               NucleiLocation, NucleationTimes_Host, NucleiGrainID, CellType, CritTimeStep, UndercoolingChange, LayerID,
-               PossibleNuclei_ThisRankThisLayer, Nuclei_WholeDomain, AtNorthBoundary, AtSouthBoundary,
-               NucleationCounter, MaxSolidificationEvents, NumberOfSolidificationEvents, LayerTimeTempHistory);
+    nucleation.placeNuclei(MaxSolidificationEvents, NumberOfSolidificationEvents, LayerTimeTempHistory, RNGSeed, 0, nx,
```
[the remark this answers is not in the document]
Comment (MattRolchigo) [reply src/runCA.cpp:226]: The problem is that `NMax`, `dTN`, and `dTsigma` are read in `InputReadFromFile`, at which point we don't have enough information to construct the Nucleation class (the Nucleation class is constructed later, once those parameters have been read and stored). My current plan is for those three variables to be part of an inputs class. Then, once the code is entirely reorganized into structs, I can have the general inputs defined in an Inputs class, with classes like Print and Nucleation inheriting the relevant inputs

File: unit_test/tstKokkosInitMPI.hpp [hunk 2]
```diff
@@ -406,33 +377,39 @@ void testNucleiInit() {
     }
 
     using memory_space = Kokkos::DefaultExecutionSpace::memory_space;
```
[the remark this answers is not in the document]
Comment (MattRolchigo) [reply unit_test/tstKokkosInitMPI.hpp:379]: using memory_space = TEST_MEMSPACE?

[the remark this answers is not in the document]
Comment (streeve) [reply unit_test/tstKokkosInitMPI.hpp:379]: Yeah that works for me

**Review by streeve — APPROVED** [review]

File: src/runCA.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (streeve) [reply src/runCA.cpp:226]: I think we're mostly on the same page here and it's fair to push this off until later

**Pull request merged**